pkg/controllers/tidbgroup/tasks/updater.go

```diff
@@ -15,138 +15,124 @@
 package tasks
 
 import (
+	"context"
 	"fmt"
+	"time"
 
 	"github.com/go-logr/logr"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 
 	"github.com/pingcap/tidb-operator/apis/core/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/action"
 	"github.com/pingcap/tidb-operator/pkg/client"
 	"github.com/pingcap/tidb-operator/pkg/runtime"
 	"github.com/pingcap/tidb-operator/pkg/updater"
 	"github.com/pingcap/tidb-operator/pkg/updater/policy"
 	"github.com/pingcap/tidb-operator/pkg/utils/k8s/revision"
 	maputil "github.com/pingcap/tidb-operator/pkg/utils/map"
 	"github.com/pingcap/tidb-operator/pkg/utils/random"
-	"github.com/pingcap/tidb-operator/pkg/utils/task"
+	"github.com/pingcap/tidb-operator/pkg/utils/task/v3"
 	"github.com/pingcap/tidb-operator/third_party/kubernetes/pkg/controller/history"
 )
 
-// TaskUpdater is a task for updating TiDBGroup when its spec is changed.
-type TaskUpdater struct {
-	Logger logr.Logger
-	Client client.Client
-	CRCli  history.Interface
-}
+const (
+	defaultUpdateWaitTime = time.Second * 30
+)
 
-func NewTaskUpdater(logger logr.Logger, c client.Client) task.Task[ReconcileContext] {
-	return &TaskUpdater{
-		Logger: logger,
-		Client: c,
-		CRCli:  history.NewClient(c),
-	}
-}
+// TaskUpdater is a task to scale or update PD when spec of TiDBGroup is changed.
+func TaskUpdater(state *ReconcileContext, c client.Client) task.Task {
+	return task.NameTaskFunc("Updater", func(ctx context.Context) task.Result {
+		logger := logr.FromContextOrDiscard(ctx)
+		historyCli := history.NewClient(c)
+		dbg := state.TiDBGroup()
+
+		selector := labels.SelectorFromSet(labels.Set{
+			// TODO(liubo02): add label of managed by operator ?
+			v1alpha1.LabelKeyCluster:   dbg.Spec.Cluster.Name,
+			v1alpha1.LabelKeyComponent: v1alpha1.LabelValComponentPD,
+			v1alpha1.LabelKeyGroup:     dbg.Name,
+		})
+
+		revisions, err := historyCli.ListControllerRevisions(dbg, selector)
```

Comment:
only TiDB Operator will create revisions for dbg, right? I think it's better to add "managed by operator"

Reply:
lgtm. I'll add it in another PR